agent/grpc-external/services/resource/delete.go

```diff
@@ -24,25 +27,26 @@ func (s *Server) Delete(ctx context.Context, req *pbresource.DeleteRequest) (*pb
 	// TODO(spatel): reg will be used for ACL hooks
 	_ = reg
 
-	if req.Version == "" {
-		// Delete resource regardless of the persisted Version. Hence, strong read
+	versionToDelete := req.Version
+	if versionToDelete == "" {
+		// Delete resource regardless of the stored Version. Hence, strong read
 		// necessary to get latest Version
 		existing, err := s.Backend.Read(ctx, storage.StrongConsistency, req.Id)
 		if err != nil {
-			return nil, err
-		}
-		if err = s.Backend.DeleteCAS(ctx, existing.Id, existing.Version); err != nil {
-			return nil, err
+			if errors.Is(err, storage.ErrNotFound) {
+				// deletes are idempotent so no-op if resource not found
+				return &pbresource.DeleteResponse{}, nil
+			}
+			return nil, fmt.Errorf("failed read: %v", err)
 		}
-		return &pbresource.DeleteResponse{}, nil
+		versionToDelete = existing.Version
 	}
 
-	// non-empty Version so let backend enforce Versions match for a CAS delete
-	if err = s.Backend.DeleteCAS(ctx, req.Id, req.Version); err != nil {
+	if err = s.Backend.DeleteCAS(ctx, req.Id, versionToDelete); err != nil {
 		if errors.Is(err, storage.ErrCASFailure) {
 			return nil, status.Error(codes.FailedPrecondition, err.Error())
```

Comment:
I've realized while reading the gRPC status [docs](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) that `Aborted` is a more appropriate code for CAS failures. 
```suggestion
			return nil, status.Error(codes.Aborted, err.Error())
```

Reply:
👍 Fixed in https://github.com/hashicorp/consul/pull/16756/commits/24714ca30e26d6ac720f02bdfe4232159f549c2c